distributed/scheduler.py

```diff
@@ -4237,104 +4238,143 @@ async def add_nanny(self) -> dict[str, Any]:
-            stripped_deps = {
-                k: v.intersection(dsk_keys)
-                for k, v in dependencies.items()
-                if k in dsk_keys
-            }
-            priority = dask.order.order(dsk, dependencies=stripped_deps)
+            graph = deserialize(graph_header, graph_frames)
+        except Exception as e:
+            err = error_message(e)
+            for key in keys:
+                self.report(
+                    {
+                        "op": "task-erred",
+                        "key": key,
+                        "exception": err["exception"],
+                        "traceback": err["traceback"],
+                    }
+                )
 
-        return self.update_graph(
-            client,
+            return
+        if isinstance(annotations, ToPickle):
+            # FIXME: why is this happening?
+            annotations = annotations.data
+        annotations = annotations or {}
+        assert isinstance(annotations, dict)
+        stimulus_id = stimulus_id or f"update-graph-{time()}"
+        (
             dsk,
-            keys,
             dependencies,
-            restrictions,
-            priority,
-            loose_restrictions,
-            resources,
-            submitting_task,
-            retries,
-            user_priority,
-            actors,
-            fifo_timeout,
-            annotations,
-            code=code,
-            stimulus_id=f"update-graph-{time()}",
-        )
+            layer_annotations,
+            pre_stringified_keys,
+        ) = self.materialize_graph(graph)
 
-    def update_graph(
-        self,
-        client=None,
-        tasks=None,
-        keys=None,
-        dependencies=None,
-        restrictions=None,
-        priority=None,
-        loose_restrictions=None,
-        resources=None,
-        submitting_task=None,
-        retries=None,
-        user_priority=0,
-        actors=None,
-        fifo_timeout=0,
-        annotations=None,
-        code=None,
-        stimulus_id=None,
-    ):
-        """
-        Add new computations to the internal dask graph
-
-        This happens whenever the Client calls submit, map, get, or compute.
-        """
-        stimulus_id = stimulus_id or f"update-graph-{time()}"
-        start = time()
-        fifo_timeout = parse_timedelta(fifo_timeout)
-        keys = set(keys)
-        if len(tasks) > 1:
+        requested_keys = set(keys)
+        del keys
+        if len(dsk) > 1:
             self.log_event(
-                ["all", client], {"action": "update_graph", "count": len(tasks)}
+                ["all", client], {"action": "update_graph", "count": len(dsk)}
             )
+        self._pop_known_tasks(dsk, dependencies)
 
-        # Remove aliases
-        for k in list(tasks):
-            if tasks[k] is k:
-                del tasks[k]
+        if lost_keys := self._pop_lost_tasks(dsk, dependencies, requested_keys):
```

Comment:
FWIW I dislike pulling non-repeated functionality into methods like this.  I find that the non-linearity it produces makes it harder for me to understand code.  This is subjective, I acknowledge.  

More thoughts here: https://matthewrocklin.com/avoid-indirection.html

Also, bonus points if you remove the walrus operator.  My guess is that most devs today don't intuitively know what it does.  It would be better, I think, to use another line and more basic syntax.

Reply:
These are just preferences, not constraints

Reply:
> FWIW I dislike pulling non-repeated functionality into methods like this. I find that the non-linearity it produces makes it harder for me to understand code. This is subjective, I acknowledge.

I tend to agree here and I was torn myself but the `update_graph` method is otherwise 400 lines of highly nontrivial code and I believe that hiding a couple of deeply nested loops and conditionals is not hurting readability. This also helps with namespace pollution.

This way I can describe what update_graph does in fewer than ten lines

- deserialize the graph
- materialize the graph
- pop known and lost keys
- generate the internal state
- apply annotations
- compute priorities
- generate recommendations

I agree there is some nuance in what "popping known and lost keys" means (or other parts) but I believe this makes the code much more accessible.

> My guess is that most devs today don't intuitively know what it does.

I think I do not agree with the statement that "most devs don't have an intuition about this". At the very least I don't think this is making accessibility to the code base any worse.
FWIW without my usage of walrus in this PR, there are already 16 occurrences of the operator in our code base